### ISSUE TITLE
What is the correct/recommended way to keep grv up-to-date when building from source ?

### ISSUE THREAD

**Issue by marcelpaulo** [rgburke/grv#56]
I've built grv from source on Xubuntu 18.04 without problems, running:

```
go get -d -v -x github.com/rgburke/grv/cmd/grv
cd $GOPATH/src/github.com/rgburke/grv
make install 
```

I'm in doubt about the correct/recommended way to keep it up-to-date. I tried:

`go get -u -v -x github.com/rgburke/grv/cmd/grv`

but it outputs `git pull` errors for all the vendor submodules, here's the full output:

```
cd /home/paulo/go/src/github.com/rgburke/grv
git config remote.origin.url
github.com/rgburke/grv (download)
cd /home/paulo/go/src/github.com/rgburke/grv
git pull --ff-only
cd /home/paulo/go/src/github.com/rgburke/grv
git submodule update --init --recursive
cd /home/paulo/go/src/github.com/rgburke/grv
git show-ref
cd /home/paulo/go/src/github.com/rgburke/grv
git submodule update --init --recursive
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/Sirupsen/logrus
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/github.com/Sirupsen/logrus (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/Sirupsen/logrus
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/Sirupsen/logrus; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/github.com/Sirupsen/logrus: exit status 1
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/bradfitz/slice
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/github.com/bradfitz/slice (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/bradfitz/slice
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/bradfitz/slice; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/github.com/bradfitz/slice: exit status 1
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/gobwas/glob
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/github.com/gobwas/glob (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/gobwas/glob
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/gobwas/glob; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/github.com/gobwas/glob: exit status 1
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/mattn/go-runewidth
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/github.com/mattn/go-runewidth (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/mattn/go-runewidth
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/mattn/go-runewidth; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/github.com/mattn/go-runewidth: exit status 1
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/rgburke/goncurses
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/github.com/rgburke/goncurses (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/rgburke/goncurses
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/rgburke/goncurses; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/github.com/rgburke/goncurses: exit status 1
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/rjeczalik/notify
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/github.com/rjeczalik/notify (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/rjeczalik/notify
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/rjeczalik/notify; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/github.com/rjeczalik/notify: exit status 1
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/tchap/go-patricia
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/github.com/tchap/go-patricia (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/tchap/go-patricia
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/github.com/tchap/go-patricia; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/github.com/tchap/go-patricia/patricia: exit status 1
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/gopkg.in/libgit2/git2go.v26
git config remote.origin.url
github.com/rgburke/grv/cmd/grv/vendor/gopkg.in/libgit2/git2go.v26 (download)
cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/gopkg.in/libgit2/git2go.v26
git pull --ff-only
# cd /home/paulo/go/src/github.com/rgburke/grv/cmd/grv/vendor/gopkg.in/libgit2/git2go.v26; git pull --ff-only
You are not currently on a branch.
Please specify which branch you want to merge with.
See git-pull(1) for details.

    git pull <remote> <branch>

package github.com/rgburke/grv/cmd/grv/vendor/gopkg.in/libgit2/git2go.v26: exit status 1
```

I've been using:

```
git pull --recurse-submodules
cd $GOPATH/src/github.com/rgburke/grv
make install 
```

but I'd like to know what is the _go-way_, so to speak, your recommended way of doing it.

**Comment by rgburke** [comment]
The command you listed above is indeed the go-way of updating a project:
```
go get -u -v -x github.com/rgburke/grv/cmd/grv
```
However there is a long standing issue with `go get -u` and git submodule vendor packages described here: golang/go#17522. So you are completely correct, however due to this issue with `go get -u` it currently won't work 😄. Apparently this will be fixed in go version 1.11.

I normally update using steps similar to yours:
```
cd $GOPATH/src/github.com/rgburke/grv
git pull
make install
```

GRV uses git submodules which are tied to a specific commit. So although doing `git pull --recurse-submodules` does update each submodule (e.g. pull's down remote branches etc...), it doesn't change HEAD in that repository and the version of the submodule that GRV is built against.
This is intended and has the advantage that everyone builds GRV using the same versions of all dependencies. The disadvantage is that we're not using the latest versions of the dependencies.

If you do want to use the latest version of each dependecy then you can run:
```
git submodule -q foreach --recursive git reset -q --hard
git submodule update --init --recursive --remote
```
This will fetch the latest master and update `HEAD` in each git submodule. However this will also show as unstaged changes when doing a `git status` in GRV's repository. You will also need to comment out the line `git submodule update --init --recursive` under the `update` target in the Makefile before building GRV otherwise it will reset each submodule back to its configured commit.

In summary, doing a `git pull` in the GRV repository is sufficient to pull down the latest changes.

**Comment by marcelpaulo** [comment]
Wow, thanks for taking the time to reply with such detail, @rgburke, and the advanced git lesson that it pulled ! When I think I got something right with git, I realize I still have so much more to learn. I've updated my build recipes.

**Comment by rgburke** [comment]
I'm glad the info was helpful @marcelpaulo. I've also found that when it comes to git I never stop learning 😄
